ofrak_core/ofrak/core/elf/load_alignment_modifier.py

```diff
@@ -0,0 +1,178 @@
+            await resource.create_child_from_view(
+                free_space,
+                data_range=Range.from_size(first_header_range.end, trimmed_potential_free_space),
+            )
+            await first_program_header.resource.run(
+                ElfProgramHeaderModifier,
+                ElfProgramHeaderModifierConfig(
+                    p_filesz=first_program_header.p_filesz + trimmed_potential_free_space,
+                    p_memsz=first_program_header.p_filesz + trimmed_potential_free_space,
+                ),
+            )
+        if found_space:
+            # Tag ELF as Allocatable so FreeSpace can be used
+            resource.add_tag(Allocatable)
+
+    @staticmethod
+    async def _trim_potential_free_space_by_vaddr(
+        first_program_header: ElfProgramHeader,
+        load_program_headers_by_offset: List[ElfProgramHeader],
+        offset_difference: int,
+    ) -> int:
+        """
+        Check that the potential free space does not overlap with any existing PT_LOAD segments.
+        If there is overlap, trim the potential free space.
+        """
+        vaddr_ranges = Range.merge_ranges(
+            [
+                Range.from_size(header.p_vaddr, header.p_memsz)
+                for header in load_program_headers_by_offset
+            ]
+        )
+        # Reduce size based on vaddr constraints
+        potential_free_space_range = Range.from_size(
+            first_program_header.p_vaddr + first_program_header.p_memsz, offset_difference
+        )
+        for vaddr_range in vaddr_ranges:
+            if potential_free_space_range.overlaps(vaddr_range):
+                potential_free_space_range = _crop_range(potential_free_space_range, vaddr_range)
+        return potential_free_space_range.length()
+
+
+async def get_load_program_headers(elf: Elf) -> List[ElfProgramHeader]:
+    load_program_headers = list()
+    for program_header in await elf.get_program_headers():
+        if program_header.p_type == ElfProgramHeaderType.LOAD.value:
+            load_program_headers.append(program_header)
+    return load_program_headers
+
+
+def _crop_range(range_1: Range, range_2: Range) -> Range:
+    """
+    Split range_1 around range_2 and return the split range whose start is the same as range_1.
+    """
+    splits = range_1.split(range_2)
+    for s in splits:
+        if s.start == range_1.start:
+            return s
+    raise ValueError(
+        f"Error removing intersection between {range_1} and {range_2} from {range_1}. "
+        f"This should be unreachable."
+    )
+
+
+def pairwise(iterable):
+    """
+    A helper function waiting for itertools.pairwise from Python 3.10:
+    https://docs.python.org/3/library/itertools.html#itertools.pairwise.
+    """
+    # pairwise('ABCDEFG') --> AB BC CD DE EF FG
+    a, b = tee(iterable)
+    next(b, None)
+    return zip(a, b)
+
+
+async def main(ofrak_context: OFRAKContext, file: str, output_file: str):
```

Comment:
IIRC when Dan tried to use this before injecting code, @whyitfor you told him that running this modifier and injecting need to happen in the same run. So is this file being runnable actually useful? You can't actually run it against a file, then open the resulting file in OFRAK and inject into it.

Reply:
Yeah best left as a separate PR. But I can see it being an Unpacker that runs on the top-level ELF, as long as it runs in parallel with the normal ELF unpacker.